packages/ai-jsx/src/batteries/sidekick/platform/conversation.tsx

```diff
@@ -77,27 +78,47 @@ export async function getNextConversationStep(
   messages: ConversationMessage[],
   fullConversation: ConversationMessage[],
   finalSystemMessageBeforeResponse: AI.Node,
-  tools: UseToolsProps['tools']
+  tools: UseToolsProps['tools'],
+  componentContext: AI.ComponentContext
 ) {
   const shrinkableConversation = getShrinkableConversation(messages, fullConversation);
   const lastMessage = messages[messages.length - 1];
+
+  // TODO: only use messages to focus on last round of conversation?
+  // Q: doess messages have all the messages from current round?
+  const updatedTools = { ...tools, ...redactedFunctionTools(fullConversation, componentContext) };
+
   switch (lastMessage.type) {
     case 'functionCall': {
       const { name, args } = lastMessage.element.props;
+      let response;
       try {
-        return <FunctionResponse name={name}>{await tools[name].func(args)}</FunctionResponse>;
+        response = await updatedTools[name].func(args);
       } catch (e: any) {
         return (
           <FunctionResponse failed name={name}>
-            {e.message}
+            <TruncateByChars maxLength={2000}>{e.message}</TruncateByChars>
```

Comment:
This might be confusing, but here `maxLength` is based on characters, but in `LargeFunctionResponseHandler` it's based on tokens by default (i.e. `lengthFunction` is set to `tokenCounter`). I wanted to allow both chars and tokens here but the effort didn't seem worth it at the time. I can take another pass at it though if need be.